@Timed and @Metered metrics are not working

`@Timed` metric is not working, RuntimeException is thrown when accessing endpoint annotated with `@Timed`.

Repo: https://github.com/rsvoboda/protean-shamrock/tree/TimedMetricTest 
Test: https://github.com/rsvoboda/protean-shamrock/commit/00b7a4ab393d05649b67107568f6b80708297902

```java
Caused by: java.lang.RuntimeException: Cannot invoke method: 
  io.smallrye.metrics.interceptors.TimedInterceptor#timedMethod on 
  io.smallrye.metrics.interceptors.TimedInterceptor@c5e89ed
```
```java
Caused by: java.lang.IllegalStateException:
  No timer with name [proccessItemTimes] found in registry 
```

```java
ERROR: UT005023: Exception handling request to /metricsresource/process
org.jboss.resteasy.spi.UnhandledException: java.lang.RuntimeException: Cannot invoke method: io.smallrye.metrics.interceptors.TimedInterceptor#timedMethod on io.smallrye.metrics.interceptors.TimedInterceptor@c5e89ed
	at org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:103)
	at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:323)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:209)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:495)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:252)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:153)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:362)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:156)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:238)
	at org.jboss.resteasy.plugins.server.servlet.ServletContainerDispatcher.service(ServletContainerDispatcher.java:229)
	at org.jboss.shamrock.resteasy.runtime.ResteasyFilter$ResteasyResponseWrapper.sendError(ResteasyFilter.java:71)
	at io.undertow.servlet.handlers.DefaultServlet.doGet(DefaultServlet.java:173)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:686)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:791)
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:74)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
	at org.jboss.shamrock.resteasy.runtime.ResteasyFilter.doFilter(ResteasyFilter.java:42)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.websockets.jsr.JsrWebSocketFilter.doFilter(JsrWebSocketFilter.java:173)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.opentracing.contrib.jaxrs2.server.SpanFinishingFilter.doFilter(SpanFinishingFilter.java:52)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:132)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AuthenticationConstraintHandler.handleRequest(AuthenticationConstraintHandler.java:53)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
	at io.undertow.servlet.handlers.security.ServletSecurityConstraintHandler.handleRequest(ServletSecurityConstraintHandler.java:59)
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:292)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:138)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:135)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at org.jboss.shamrock.undertow.runtime.UndertowDeploymentTemplate$7$1$1.call(UndertowDeploymentTemplate.java:358)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:272)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:104)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:364)
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1998)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1525)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1382)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: Cannot invoke method: io.smallrye.metrics.interceptors.TimedInterceptor#timedMethod on io.smallrye.metrics.interceptors.TimedInterceptor@c5e89ed
	at org.jboss.protean.arc.Reflections.invokeMethod(Reflections.java:101)
	at io.smallrye.metrics.interceptors.TimedInterceptor_Bean.intercept(Unknown Source)
	at org.jboss.protean.arc.InvocationContextImpl$InterceptorInvocation.invoke(InvocationContextImpl.java:263)
	at org.jboss.protean.arc.InvocationContextImpl.invokeNext(InvocationContextImpl.java:143)
	at org.jboss.protean.arc.InvocationContextImpl.proceed(InvocationContextImpl.java:167)
	at org.jboss.shamrock.example.metrics.MetricsResource_Subclass.proccessItem(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:152)
	at org.jboss.resteasy.core.MethodInjectorImpl.lambda$invoke$3(MethodInjectorImpl.java:123)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
	at java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:614)
	at java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:1983)
	at java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:110)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:123)
	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:543)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:418)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:372)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:362)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:374)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:343)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invoke$1(ResourceMethodInvoker.java:317)
	at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:981)
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2124)
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:110)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:317)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:476)
	... 54 more
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.protean.arc.Reflections.invokeMethod(Reflections.java:99)
	... 82 more
Caused by: java.lang.IllegalStateException: No timer with name [proccessItemTimes] found in registry [io.smallrye.metrics.MetricRegistries_ProducerMethod_getApplicationRegistry_3084f885749d4d534e6c1a0434dc45444ee9cbec_ClientProxy@6dd208f5]
	at io.smallrye.metrics.interceptors.TimedInterceptor.timedCallable(TimedInterceptor.java:75)
	at io.smallrye.metrics.interceptors.TimedInterceptor.timedMethod(TimedInterceptor.java:63)
	... 87 more
```

--- a/extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java
+++ b/extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java
@@ -19,20 +19,19 @@ package io.quarkus.smallrye.metrics.deployment;
 import static io.quarkus.deployment.annotations.ExecutionTime.RUNTIME_INIT;
 import static io.quarkus.deployment.annotations.ExecutionTime.STATIC_INIT;
 
+import java.util.Arrays;
 import java.util.Collection;
 
-import javax.interceptor.Interceptor;
-
 import org.eclipse.microprofile.metrics.annotation.Counted;
+import org.eclipse.microprofile.metrics.annotation.Gauge;
+import org.eclipse.microprofile.metrics.annotation.Metered;
+import org.eclipse.microprofile.metrics.annotation.Timed;
 import org.jboss.jandex.AnnotationInstance;
-import org.jboss.jandex.AnnotationTarget;
-import org.jboss.jandex.AnnotationTarget.Kind;
-import org.jboss.jandex.ClassInfo;
+import org.jboss.jandex.AnnotationValue;
 import org.jboss.jandex.DotName;
-import org.jboss.jandex.IndexView;
-import org.jboss.jandex.MethodInfo;
 
 import io.quarkus.arc.deployment.AdditionalBeanBuildItem;
+import io.quarkus.arc.deployment.AnnotationsTransformerBuildItem;
 import io.quarkus.arc.deployment.BeanArchiveIndexBuildItem;
 import io.quarkus.arc.deployment.BeanContainerBuildItem;
 import io.quarkus.deployment.annotations.BuildProducer;
@@ -59,6 +58,12 @@ import io.smallrye.metrics.interceptors.TimedInterceptor;
 public class SmallRyeMetricsProcessor {
 
     SmallRyeMetricsConfig metrics;
+
+    private static final Collection<DotName> metricsAnnotations = Arrays.asList(
+            DotName.createSimple(Gauge.class.getName()),
+            DotName.createSimple(Counted.class.getName()),
+            DotName.createSimple(Timed.class.getName()),
+            DotName.createSimple(Metered.class.getName()));
 
     @ConfigRoot(name = "smallrye-metrics")
     static final class SmallRyeMetricsConfig {
@@ -92,44 +97,44 @@ public class SmallRyeMetricsProcessor {
     }
 
     @BuildStep
+    void annotationTransformers(BuildProducer<AnnotationsTransformerBuildItem> transformers) {
+        // attach @MetricsBinding to each class that contains any metric annotations
+        transformers.produce(new AnnotationsTransformerBuildItem(ctx -> {
+            if (ctx.isClass()) {
+                // skip classes in package io.smallrye.metrics.interceptors
+                if (ctx.getTarget().asClass().name().toString()
+                        .startsWith(io.smallrye.metrics.interceptors.MetricsInterceptor.class.getPackage().getName())) {
+                    return;
+                }
+                for (DotName annotationName : ctx.getTarget().asClass().annotations().keySet()) {
+                    if (metricsAnnotations.contains(annotationName)) {
+                        ctx.transform().add(AnnotationInstance.create(DotName.createSimple(MetricsBinding.class.getName()),
+                                ctx.getTarget(), new AnnotationValue[0]))
+                                .done();
+                        return;
+                    }
+                }
+            }
+        }));
+    }
+
+    @BuildStep
     @Record(STATIC_INIT)
     public void build(BeanContainerBuildItem beanContainerBuildItem,
             SmallRyeMetricsTemplate metrics,
             ShutdownContextBuildItem shutdown,
             BuildProducer<ReflectiveClassBuildItem> reflectiveClasses,
             BeanArchiveIndexBuildItem beanArchiveIndex,
-            BuildProducer<FeatureBuildItem> feature) throws Exception {
+            BuildProducer<FeatureBuildItem> feature) {
 
         feature.produce(new FeatureBuildItem(FeatureBuildItem.SMALLRYE_METRICS));
 
-        reflectiveClasses.produce(new ReflectiveClassBuildItem(false, false, Counted.class.getName()));
+        for (DotName metricsAnnotation : metricsAnnotations) {
+            reflectiveClasses.produce(new ReflectiveClassBuildItem(false, false, metricsAnnotation.toString()));
+        }
         reflectiveClasses.produce(new ReflectiveClassBuildItem(false, false, MetricsBinding.class.getName()));
 
         metrics.createRegistries(beanContainerBuildItem.getValue());
-
-        IndexView index = beanArchiveIndex.getIndex();
-        Collection<AnnotationInstance> annos = index.getAnnotations(DotName.createSimple(Counted.class.getName()));
-
-        for (AnnotationInstance anno : annos) {
-            AnnotationTarget target = anno.target();
-
-            // We need to exclude metrics interceptors
-            if (Kind.CLASS.equals(target.kind())
-                    && target.asClass().classAnnotations().stream()
-                            .anyMatch(a -> a.name().equals(DotName.createSimple(Interceptor.class.getName())))) {
-                continue;
-            }
-
-            MethodInfo methodInfo = target.asMethod();
-            String name = methodInfo.name();
-            if (anno.value("name") != null) {
-                name = anno.value("name").asString();
-            }
-            ClassInfo classInfo = methodInfo.declaringClass();
-
-            metrics.registerCounted(classInfo.name().toString(),
-                    name, shutdown);
-        }
     }
 
     @BuildStep

--- a/extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java
+++ b/extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java
@@ -20,7 +20,8 @@ import static io.quarkus.deployment.annotations.ExecutionTime.RUNTIME_INIT;
 import static io.quarkus.deployment.annotations.ExecutionTime.STATIC_INIT;
 
 import java.util.Arrays;
-import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
 
 import org.eclipse.microprofile.metrics.annotation.Counted;
 import org.eclipse.microprofile.metrics.annotation.Gauge;
@@ -57,13 +58,11 @@ import io.smallrye.metrics.interceptors.TimedInterceptor;
 
 public class SmallRyeMetricsProcessor {
 
-    SmallRyeMetricsConfig metrics;
-
-    private static final Collection<DotName> metricsAnnotations = Arrays.asList(
+    private static final Set<DotName> metricsAnnotations = new HashSet<>(Arrays.asList(
             DotName.createSimple(Gauge.class.getName()),
             DotName.createSimple(Counted.class.getName()),
             DotName.createSimple(Timed.class.getName()),
-            DotName.createSimple(Metered.class.getName()));
+            DotName.createSimple(Metered.class.getName())));
 
     @ConfigRoot(name = "smallrye-metrics")
     static final class SmallRyeMetricsConfig {
@@ -74,6 +73,8 @@ public class SmallRyeMetricsProcessor {
         @ConfigItem(defaultValue = "/metrics")
         String path;
     }
+
+    SmallRyeMetricsConfig metrics;
 
     @BuildStep
     ServletBuildItem createServlet() {

--- a/extensions/smallrye-metrics/runtime/src/main/java/io/quarkus/smallrye/metrics/runtime/SmallRyeMetricsTemplate.java
+++ b/extensions/smallrye-metrics/runtime/src/main/java/io/quarkus/smallrye/metrics/runtime/SmallRyeMetricsTemplate.java
@@ -32,7 +32,6 @@ import io.quarkus.arc.runtime.BeanContainer;
 import io.quarkus.runtime.ShutdownContext;
 import io.quarkus.runtime.annotations.Template;
 import io.smallrye.metrics.MetricRegistries;
-import io.smallrye.metrics.app.CounterImpl;
 
 @Template
 public class SmallRyeMetricsTemplate {
@@ -41,31 +40,6 @@ public class SmallRyeMetricsTemplate {
     private static final String MEMORY_HEAP_USAGE = "memory.heap.usage";
     private static final String MEMORY_NON_HEAP_USAGE = "memory.nonHeap.usage";
     private static final String THREAD_COUNT = "thread.count";
-
-    /*
-     * public <E extends Member & AnnotatedElement> void registerCounted(Class<?> topClass, E element) {
-     * MetricRegistry registry = MetricRegistries.get(MetricRegistry.Type.APPLICATION);
-     * //registry.register(name, new CounterImpl());
-     * MetricResolver resolver = new MetricResolver();
-     * MetricResolver.Of<Counted> of = resolver.counted(topClass, element);
-     * Metadata meta = new Metadata(of.metricName(), MetricType.COUNTER);
-     * //registry.register(meta, new CounterImpl());
-     * }
-     */
-    public void registerCounted(String topClassName, String elementName, ShutdownContext shutdown) {
-        MetricRegistry registry = MetricRegistries.get(MetricRegistry.Type.APPLICATION);
-
-        String name = MetricRegistry.name(topClassName, elementName);
-        Metadata meta = new Metadata(name, MetricType.COUNTER);
-        log.debugf("Register: %s", name);
-        registry.register(meta, new CounterImpl());
-        shutdown.addShutdownTask(new Runnable() {
-            @Override
-            public void run() {
-                registry.remove(name);
-            }
-        });
-    }
 
     public void registerBaseMetrics(ShutdownContext shutdown) {
         MetricRegistry registry = MetricRegistries.get(MetricRegistry.Type.BASE);

--- a/integration-tests/main/src/main/java/io/quarkus/example/metrics/MetricsResource.java
+++ b/integration-tests/main/src/main/java/io/quarkus/example/metrics/MetricsResource.java
@@ -16,17 +16,71 @@
 
 package io.quarkus.example.metrics;
 
+import javax.inject.Inject;
 import javax.ws.rs.GET;
 import javax.ws.rs.Path;
 
+import org.eclipse.microprofile.metrics.Histogram;
+import org.eclipse.microprofile.metrics.MetricUnits;
 import org.eclipse.microprofile.metrics.annotation.Counted;
+import org.eclipse.microprofile.metrics.annotation.Gauge;
+import org.eclipse.microprofile.metrics.annotation.Metered;
+import org.eclipse.microprofile.metrics.annotation.Metric;
+import org.eclipse.microprofile.metrics.annotation.Timed;
 
 @Path("/metricsresource")
 public class MetricsResource {
 
+    @Inject
+    @Metric(name = "histogram")
+    Histogram histogram;
+
     @GET
+    @Path("/counter")
     @Counted(monotonic = true, name = "a_counted_resource")
-    public String getTest() {
+    public String counter() {
+        return "TEST";
+    }
+
+    @GET
+    @Path("/gauge")
+    @Gauge(name = "gauge", unit = MetricUnits.NONE)
+    public Long gauge() {
+        return 42L;
+    }
+
+    @GET
+    @Path("/meter")
+    @Metered(name = "meter")
+    public String meter() {
+        return "OK";
+    }
+
+    @GET
+    @Path("/timer")
+    @Timed(name = "timer_metric")
+    public String timer() {
+        return "OK";
+    }
+
+    @GET
+    @Path("/histogram")
+    public String histogram() {
+        histogram.update(42);
+        return "OK";
+    }
+
+    @GET
+    @Path("/counter-absolute")
+    @Counted(monotonic = true, name = "counter_absolute", absolute = true)
+    public String counterAbsoluteName() {
+        return "TEST";
+    }
+
+    @GET
+    @Path("/counter-with-tags")
+    @Counted(monotonic = true, name = "counter_with_tags", tags = "foo=bar")
+    public String counterWithTags() {
         return "TEST";
     }
 

--- a/integration-tests/main/src/test/java/io/quarkus/example/test/MetricsTestCase.java
+++ b/integration-tests/main/src/test/java/io/quarkus/example/test/MetricsTestCase.java
@@ -28,10 +28,60 @@ import io.restassured.RestAssured;
 public class MetricsTestCase {
 
     @Test
-    public void testMetrics() {
-        testCounted("0.0");
-        invokeResource();
-        testCounted("1.0");
+    public void testCounter() {
+        invokeCounter();
+        assertMetricExactValue("application:io_quarkus_example_metrics_metrics_resource_a_counted_resource", "1.0");
+    }
+
+    @Test
+    public void testGauge() {
+        invokeGauge();
+        assertMetricExactValue("application:io_quarkus_example_metrics_metrics_resource_gauge", "42.0");
+    }
+
+    @Test
+    public void testMeter() {
+        invokeMeter();
+        assertMetricExactValue("application:io_quarkus_example_metrics_metrics_resource_meter_total", "1.0");
+        assertMetricExists("application:io_quarkus_example_metrics_metrics_resource_meter_rate_per_second");
+        assertMetricExists("application:io_quarkus_example_metrics_metrics_resource_meter_one_min_rate_per_second");
+        assertMetricExists("application:io_quarkus_example_metrics_metrics_resource_meter_five_min_rate_per_second");
+        assertMetricExists("application:io_quarkus_example_metrics_metrics_resource_meter_fifteen_min_rate_per_second");
+    }
+
+    @Test
+    public void testTimer() {
+        invokeTimer();
+        assertMetricExactValue("application:io_quarkus_example_metrics_metrics_resource_timer_metric_seconds_count", "1.0");
+        assertMetricExists("application:io_quarkus_example_metrics_metrics_resource_timer_metric_seconds{quantile=\"0.5\"}");
+        assertMetricExists("application:io_quarkus_example_metrics_metrics_resource_timer_metric_seconds{quantile=\"0.75\"}");
+        assertMetricExists("application:io_quarkus_example_metrics_metrics_resource_timer_metric_seconds{quantile=\"0.95\"}");
+        assertMetricExists("application:io_quarkus_example_metrics_metrics_resource_timer_metric_seconds{quantile=\"0.98\"}");
+        assertMetricExists("application:io_quarkus_example_metrics_metrics_resource_timer_metric_seconds{quantile=\"0.99\"}");
+        assertMetricExists("application:io_quarkus_example_metrics_metrics_resource_timer_metric_seconds{quantile=\"0.999\"}");
+        assertMetricExists("application:io_quarkus_example_metrics_metrics_resource_timer_metric_one_min_rate_per_second");
+        assertMetricExists("application:io_quarkus_example_metrics_metrics_resource_timer_metric_five_min_rate_per_second");
+        assertMetricExists("application:io_quarkus_example_metrics_metrics_resource_timer_metric_fifteen_min_rate_per_second");
+        assertMetricExists("application:io_quarkus_example_metrics_metrics_resource_timer_metric_min_seconds");
+        assertMetricExists("application:io_quarkus_example_metrics_metrics_resource_timer_metric_max_seconds");
+        assertMetricExists("application:io_quarkus_example_metrics_metrics_resource_timer_metric_mean_seconds");
+        assertMetricExists("application:io_quarkus_example_metrics_metrics_resource_timer_metric_stddev_seconds");
+    }
+
+    @Test
+    public void testHistogram() {
+        invokeHistogram();
+        assertMetricExactValue("application:io_quarkus_example_metrics_metrics_resource_histogram_count", "1.0");
+        assertMetricExactValue("application:io_quarkus_example_metrics_metrics_resource_histogram_min", "42.0");
+        assertMetricExactValue("application:io_quarkus_example_metrics_metrics_resource_histogram_max", "42.0");
+        assertMetricExactValue("application:io_quarkus_example_metrics_metrics_resource_histogram_mean", "42.0");
+        assertMetricExactValue("application:io_quarkus_example_metrics_metrics_resource_histogram_stddev", "0.0");
+        assertMetricExactValue("application:io_quarkus_example_metrics_metrics_resource_histogram{quantile=\"0.5\"}", "42.0");
+        assertMetricExactValue("application:io_quarkus_example_metrics_metrics_resource_histogram{quantile=\"0.75\"}", "42.0");
+        assertMetricExactValue("application:io_quarkus_example_metrics_metrics_resource_histogram{quantile=\"0.95\"}", "42.0");
+        assertMetricExactValue("application:io_quarkus_example_metrics_metrics_resource_histogram{quantile=\"0.98\"}", "42.0");
+        assertMetricExactValue("application:io_quarkus_example_metrics_metrics_resource_histogram{quantile=\"0.99\"}", "42.0");
+        assertMetricExactValue("application:io_quarkus_example_metrics_metrics_resource_histogram{quantile=\"0.999\"}", "42.0");
     }
 
     @Test
@@ -43,6 +93,18 @@ public class MetricsTestCase {
     }
 
     @Test
+    public void testMetricWithAbsoluteName() {
+        invokeCounterWithAbsoluteName();
+        assertMetricExactValue("application:counter_absolute", "1.0");
+    }
+
+    @Test
+    public void testMetricWithCustomTags() {
+        invokeCounterWithTags();
+        assertMetricExactValue("application:io_quarkus_example_metrics_metrics_resource_counter_with_tags{foo=\"bar\"}", "1.0");
+    }
+
+    @Test
     public void testInvalidScopes() {
         RestAssured.when().get("/metrics/foo").then().statusCode(404)
                 .body(containsString("Bad scope requested"));
@@ -50,13 +112,48 @@ public class MetricsTestCase {
                 .body(containsString("Metric vendor/foo not found"));
     }
 
-    private void testCounted(String val) {
+    private void assertMetricExactValue(String name, String val) {
         RestAssured.when().get("/metrics").then()
-                .body(containsString("application:io_quarkus_example_metrics_metrics_resource_a_counted_resource " + val));
+                .body(containsString(name + " " + val));
     }
 
-    public void invokeResource() {
-        RestAssured.when().get("/metricsresource").then()
+    private void assertMetricExists(String name) {
+        RestAssured.when().get("/metrics").then()
+                .body(containsString(name));
+    }
+
+    public void invokeCounter() {
+        RestAssured.when().get("/metricsresource/counter").then()
+                .body(is("TEST"));
+    }
+
+    public void invokeGauge() {
+        RestAssured.when().get("/metricsresource/gauge").then()
+                .body(is("42"));
+    }
+
+    public void invokeMeter() {
+        RestAssured.when().get("/metricsresource/meter").then()
+                .body(is("OK"));
+    }
+
+    public void invokeTimer() {
+        RestAssured.when().get("/metricsresource/timer").then()
+                .body(is("OK"));
+    }
+
+    public void invokeHistogram() {
+        RestAssured.when().get("/metricsresource/histogram").then()
+                .body(is("OK"));
+    }
+
+    public void invokeCounterWithAbsoluteName() {
+        RestAssured.when().get("/metricsresource/counter-absolute").then()
+                .body(is("TEST"));
+    }
+
+    public void invokeCounterWithTags() {
+        RestAssured.when().get("/metricsresource/counter-with-tags").then()
                 .body(is("TEST"));
     }
 

